File: backend/app/ingest.py
from __future__ import annotations

import csv
import io
from datetime import date, datetime
from typing import Iterable

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.basket import FARE_CLASS
from app.cleaning.pipeline import clean_quotes
from app.collectors.safeguards import CollectionEvent
from app.index.construct import construct_index
from app.models import QuoteRaw
from app.schemas import QuoteIn


def quote_in_to_event(q: QuoteIn) -> CollectionEvent:
    collected_on = q.collected_on or date.today()
    lead = q.lead_time_days
    if lead is None:
        lead = (q.dep_date - collected_on).days
    return CollectionEvent(
        source=(q.source or "manual").strip() or "manual",
        origin=q.origin.strip().upper(),
        destination=q.destination.strip().upper(),
        carrier=q.carrier.strip().upper(),
        flight_no=(q.flight_no or "NA").strip() or "NA",
        dep_date=q.dep_date,
        fare_class=(q.fare_class or FARE_CLASS).strip() or FARE_CLASS,
        lead_time_days=int(lead),
        collected_on=collected_on,
        collected_at=datetime.utcnow(),
        status=(q.status or "ok").strip() or "ok",
        base_fare=q.base_fare,
        taxes=q.taxes,
        udf=q.udf,
        convenience=q.convenience,
        total_fare=q.total_fare,
    )


def _raw_match(session: Session, ev: CollectionEvent) -> QuoteRaw | None:
    return session.scalars(
        select(QuoteRaw).where(
            QuoteRaw.source == ev.source,
            QuoteRaw.origin == ev.origin,
            QuoteRaw.destination == ev.destination,
            QuoteRaw.carrier == ev.carrier,
            QuoteRaw.flight_no == ev.flight_no,
            QuoteRaw.dep_date == ev.dep_date,
            QuoteRaw.fare_class == ev.fare_class,
            QuoteRaw.collected_on == ev.collected_on,
            QuoteRaw.lead_time_days == ev.lead_time_days,
        )
    ).first()


def upsert_events(session: Session, source: str, events: list[CollectionEvent], run_id: int | None = None) -> dict[str, int]:
    counts = {"inserted": 0, "updated": 0, "ok": 0, "missing": 0, "sold_out": 0, "blocked": 0, "cancelled": 0}
    for ev in events:
        counts[ev.status] = counts.get(ev.status, 0) + 1
        existing = _raw_match(session, ev)
        if existing is None:
            session.add(
                QuoteRaw(
                    run_id=run_id,
                    source=ev.source,
                    origin=ev.origin,
                    destination=ev.destination,
                    carrier=ev.carrier or "NA",
                    flight_no=ev.flight_no or "NA",
                    dep_date=ev.dep_date,
                    fare_class=ev.fare_class,
                    lead_time_days=ev.lead_time_days,
                    collected_on=ev.collected_on,
                    collected_at=ev.collected_at,
                    status=ev.status,
                    base_fare=ev.base_fare,
                    taxes=ev.taxes,
                    udf=ev.udf,
                    convenience=ev.convenience,
                    total_fare=ev.total_fare,
                )
            )
            counts["inserted"] += 1
        else:
            existing.status = ev.status
            existing.base_fare = ev.base_fare
            existing.taxes = ev.taxes
            existing.udf = ev.udf
            existing.convenience = ev.convenience
            existing.total_fare = ev.total_fare
            existing.collected_at = ev.collected_at
            if run_id is not None:
                existing.run_id = run_id
            counts["updated"] += 1
    session.flush()
    return counts


def ingest_quotes(session: Session, quotes: Iterable[QuoteIn], *, rebuild_index: bool = True) -> dict:
    from datetime import datetime as dt

    from app.models import CollectionRun

    events = [quote_in_to_event(q) for q in quotes]
    run = CollectionRun(
        started_at=dt.utcnow(),
        source="manual",
        status="running",
    )
    session.add(run)
    session.flush()
    counts = upsert_events(session, "manual", events, run_id=run.id)
    run.finished_at = dt.utcnow()
    run.status = "ok"
    run.quotes_ok = counts.get("ok", 0)
    run.quotes_missing = counts.get("missing", 0)
    run.quotes_sold_out = counts.get("sold_out", 0)
    run.quotes_blocked = counts.get("blocked", 0)
    run.notes = f"inserted={counts['inserted']} updated={counts['updated']}"
    cleaned = 0
    indexed = 0
    if rebuild_index:
        cleaned = clean_quotes(session)
        indexed = construct_index(session)
    session.commit()
    return {
        "source": "manual",
        "received": len(events),
        **counts,
        "cleaned": cleaned,
        "index_rows": indexed,
    }


def _cell(row: dict, *names: str) -> str:
    lower = {k.strip().lower(): v for k, v in row.items() if k}
    for name in names:
        if name in lower and lower[name] not in (None, ""):
            return str(lower[name]).strip()
    return ""


def _opt_float(text: str) -> float | None:
    if text is None or str(text).strip() == "":
        return None
    return float(str(text).replace(",", "").replace("₹", "").strip())


def parse_csv_quotes(text: str) -> list[QuoteIn]:
    reader = csv.DictReader(io.StringIO(text))
    out: list[QuoteIn] = []
    for row in reader:
        origin = _cell(row, "origin", "from").upper()
        dest = _cell(row, "destination", "dest", "to").upper()
        if not origin or not dest:
            continue
        collected = _cell(row, "collected_on", "collected", "quote_date") or date.today().isoformat()
        dep = _cell(row, "dep_date", "departure", "travel_date")
        if not dep:
            continue
        lead_raw = _cell(row, "lead_time_days", "lead_time", "tplus")
        payload = {
            "source": _cell(row, "source") or "manual",
            "origin": origin,
            "destination": dest,
            "carrier": (_cell(row, "carrier", "airline") or "NA").upper(),
            "flight_no": _cell(row, "flight_no", "flight") or "NA",
            "dep_date": dep,
            "fare_class": _cell(row, "fare_class", "cabin") or FARE_CLASS,
            "collected_on": collected,
            "base_fare": _opt_float(_cell(row, "base_fare", "base")),
            "taxes": _opt_float(_cell(row, "taxes", "tax")),
            "udf": _opt_float(_cell(row, "udf")),
            "convenience": _opt_float(_cell(row, "convenience", "conv_fee")),
            "total_fare": _opt_float(_cell(row, "total_fare", "total", "fare")),
            "status": _cell(row, "status") or "ok",
        }
        if lead_raw:
            payload["lead_time_days"] = int(float(lead_raw))
        out.append(QuoteIn.model_validate(payload))
    return out

File: lib/ingest.ts
import { constructIndex } from "./apix";
import { FARE_CLASS } from "./bootstrap";
import { cleanQuotes } from "./cleaning";
import type { sql as sqlFn } from "./db";
import { isoDate } from "./db";

export type QuoteIn = {
  source?: string;
  origin: string;
  destination: string;
  carrier: string;
  flight_no?: string;
  dep_date: string;
  return_date?: string | null;
  trip_type?: string;
  fare_class?: string;
  lead_time_days?: number | null;
  collected_on?: string | null;
  base_fare?: number | null;
  taxes?: number | null;
  udf?: number | null;
  convenience?: number | null;
  total_fare?: number | null;
  status?: string;
};

export type TripType = "one_way" | "round_trip";

export function normalizeTripType(raw?: string | null): TripType {
  const t = (raw || "").toLowerCase().replace(/[\s-]+/g, "_");
  if (t === "round_trip" || t === "roundtrip" || t === "return" || t === "rt") return "round_trip";
  return "one_way";
}

export function uniqueFlightNo(flightNo: string | undefined, trip: TripType): string {
  const fn = ((flightNo || "NA").trim() || "NA").replace(/^RT\//, "");
  if (trip === "round_trip") return `RT/${fn}`.slice(0, 16);
  return fn.slice(0, 16);
}

export function displayFlightNo(flightNo: unknown): string {
  return String(flightNo ?? "").replace(/^RT\//, "") || "NA";
}

export type CollectionEvent = QuoteIn & {
  source: string;
  flight_no: string;
  fare_class: string;
  lead_time_days: number;
  collected_on: string;
  collected_at: string;
  status: string;
  trip_type: TripType;
};

function toEvent(q: QuoteIn): CollectionEvent {
  const collected_on = q.collected_on || new Date().toISOString().slice(0, 10);
  let lead = q.lead_time_days;
  if (lead == null) {
    const dep = new Date(`${q.dep_date}T00:00:00Z`);
    const col = new Date(`${collected_on}T00:00:00Z`);
    lead = Math.round((dep.getTime() - col.getTime()) / 86400000);
  }
  const trip_type = normalizeTripType(q.trip_type);
  return {
    ...q,
    source: (q.source || "manual").trim() || "manual",
    origin: q.origin.trim().toUpperCase(),
    destination: q.destination.trim().toUpperCase(),
    carrier: q.carrier.trim().toUpperCase(),
    flight_no: uniqueFlightNo(q.flight_no, trip_type),
    fare_class: (q.fare_class || FARE_CLASS).replace(/_RT$/i, "").trim() || FARE_CLASS,
    lead_time_days: Number(lead),
    collected_on,
    collected_at: new Date().toISOString(),
    status: (q.status || "ok").trim() || "ok",
    trip_type,
    return_date: q.return_date ? isoDate(q.return_date) : null,
  };
}

export async function upsertEvents(
  q: ReturnType<typeof sqlFn>,
  events: CollectionEvent[],
  runId: number | null,
): Promise<Record<string, number>> {
  const counts: Record<string, number> = {
    inserted: 0,
    updated: 0,
    ok: 0,
    missing: 0,
    sold_out: 0,
    blocked: 0,
    cancelled: 0,
  };
  for (const ev of events) {
    counts[ev.status] = (counts[ev.status] ?? 0) + 1;
    const existing = await q`
      SELECT id FROM quotes_raw
      WHERE source = ${ev.source} AND origin = ${ev.origin} AND destination = ${ev.destination}
        AND carrier = ${ev.carrier} AND flight_no = ${ev.flight_no} AND dep_date = ${isoDate(ev.dep_date)}
        AND fare_class = ${ev.fare_class} AND collected_on = ${isoDate(ev.collected_on)}
        AND lead_time_days = ${ev.lead_time_days}
      LIMIT 1
    `;
    if (!existing.length) {
      await q`
        INSERT INTO quotes_raw (
          run_id, source, origin, destination, carrier, flight_no, dep_date, fare_class,
          lead_time_days, collected_on, collected_at, status, base_fare, taxes, udf, convenience, total_fare, currency,
          trip_type, return_date
        ) VALUES (
          ${runId}, ${ev.source}, ${ev.origin}, ${ev.destination}, ${ev.carrier}, ${ev.flight_no},
          ${isoDate(ev.dep_date)}, ${ev.fare_class}, ${ev.lead_time_days}, ${isoDate(ev.collected_on)}, ${ev.collected_at},
          ${ev.status}, ${ev.base_fare ?? null}, ${ev.taxes ?? null}, ${ev.udf ?? null},
          ${ev.convenience ?? null}, ${ev.total_fare ?? null}, 'INR',
          ${ev.trip_type}, ${ev.return_date ?? null}
        )
      `;
      counts.inserted += 1;
    } else {
      await q`
        UPDATE quotes_raw SET
          status = ${ev.status},
          base_fare = ${ev.base_fare ?? null},
          taxes = ${ev.taxes ?? null},
          udf = ${ev.udf ?? null},
          convenience = ${ev.convenience ?? null},
          total_fare = ${ev.total_fare ?? null},
          collected_at = ${ev.collected_at},
          run_id = ${runId},
          trip_type = ${ev.trip_type},
          return_date = ${ev.return_date ?? null}
        WHERE id = ${existing[0].id}
      `;
      counts.updated += 1;
    }
  }
  return counts;
}

export async function ingestQuotes(
  q: ReturnType<typeof sqlFn>,
  quotes: QuoteIn[],
  rebuildIndex = true,
): Promise<Record<string, unknown>> {
  const events = quotes.map(toEvent);
  const started = new Date().toISOString();
  const run = await q`
    INSERT INTO collection_runs (started_at, source, status, quotes_ok, quotes_missing, quotes_sold_out, quotes_blocked, notes)
    VALUES (${started}, 'manual', 'running', 0, 0, 0, 0, '')
    RETURNING id
  `;
  const runId = Number(run[0].id);
  const counts = await upsertEvents(q, events, runId);
  let cleaned = 0;
  let indexed = 0;
  if (rebuildIndex) {
    cleaned = await cleanQuotes(q);
    indexed = await constructIndex(q);
  }
  const finished = new Date().toISOString();
  await q`
    UPDATE collection_runs SET
      finished_at = ${finished},
      status = 'ok',
      quotes_ok = ${counts.ok ?? 0},
      quotes_missing = ${counts.missing ?? 0},
      quotes_sold_out = ${counts.sold_out ?? 0},
      quotes_blocked = ${counts.blocked ?? 0},
      notes = ${`inserted=${counts.inserted} updated=${counts.updated}`}
    WHERE id = ${runId}
  `;
  return { source: "manual", received: events.length, ...counts, cleaned, index_rows: indexed };
}

function cell(row: Record<string, string>, ...names: string[]): string {
  const lower: Record<string, string> = {};
  for (const [k, v] of Object.entries(row)) {
    if (k) lower[k.trim().toLowerCase()] = v;
  }
  for (const name of names) {
    if (lower[name] != null && lower[name] !== "") return String(lower[name]).trim();
  }
  return "";
}

function optFloat(text: string): number | null {
  if (!text.trim()) return null;
  return Number(text.replace(/,/g, "").replace(/₹/g, "").trim());
}

export function parseCsvQuotes(text: string): QuoteIn[] {
  const lines = text.replace(/^\uFEFF/, "").trim().split(/\r?\n/);
  if (lines.length < 2) return [];
  const headers = lines[0].split(",").map((h) => h.trim());
  const out: QuoteIn[] = [];
  for (const line of lines.slice(1)) {
    if (!line.trim()) continue;
    const cols = line.split(",");
    const row: Record<string, string> = {};
    headers.forEach((h, i) => {
      row[h] = cols[i] ?? "";
    });
    const origin = cell(row, "origin", "from").toUpperCase();
    const dest = cell(row, "destination", "dest", "to").toUpperCase();
    if (!origin || !dest) continue;
    const dep = cell(row, "dep_date", "departure", "travel_date");
    if (!dep) continue;
    const leadRaw = cell(row, "lead_time_days", "lead_time", "tplus");
    const trip = normalizeTripType(cell(row, "trip_type", "trip", "journey"));
    const q: QuoteIn = {
      source: cell(row, "source") || "manual",
      origin,
      destination: dest,
      carrier: (cell(row, "carrier", "airline") || "NA").toUpperCase(),
      flight_no: cell(row, "flight_no", "flight") || "NA",
      dep_date: dep,
      return_date: cell(row, "return_date", "return") || null,
      trip_type: trip,
      fare_class: cell(row, "fare_class", "cabin") || FARE_CLASS,
      collected_on: cell(row, "collected_on", "collected", "quote_date") || new Date().toISOString().slice(0, 10),
      base_fare: optFloat(cell(row, "base_fare", "base")),
      taxes: optFloat(cell(row, "taxes", "tax")),
      udf: optFloat(cell(row, "udf")),
      convenience: optFloat(cell(row, "convenience", "conv_fee")),
      total_fare: optFloat(cell(row, "total_fare", "total", "fare")),
      status: cell(row, "status") || "ok",
    };
    if (leadRaw) q.lead_time_days = Number(leadRaw);
    out.push(q);
  }
  return out;
}
